streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValues.java

```diff
@@ -46,8 +58,104 @@ public KStreamTransformValuesProcessor(ValueTransformer<V, R> valueTransformer)
         }
 
         @Override
-        public void init(ProcessorContext context) {
-            valueTransformer.init(context);
+        public void init(final ProcessorContext context) {
```

Comment:
I think I remembered the original motivation for allowing users to return not-null values in `punctuate` now, which is that users may have a n-to-m mapping in the transform, and we do not have a `flatTransform` or `flatTransformValues` yet (we have a JIRA open for it); and even if we have this function, it may be inefficient as some scenarios cannot determine the return value(s) just from the single input, but may from multiple consecutive values (i.e. it is `n-to-m` not `1-to-m` or `1-to-1`). That is why we modified the punctuate behavior earlier to return some value.

But as @mjsax observed, returning `<null, R returnValue>` would not be appropriate. Instead I think we should set the key as the latest observed incoming record's key (i.e. we should remember that key).

As for `context.forward()`, I think throwing exceptions is the right thing to do; unfortunately we need to implement it as a wrapper. I think ideally we should separate the following into different interfaces:

```

ProcessorContext: applicationId, taskId, key / value serdes, metrics, get / register stores, commit, schedule, forward

RecordContext: timestamp, topic, partition, offset
```

And in the DSL's `process() / transform()` as well as rich functions that we are considering to add, we only expose `RecordContext`. 

If people agree on this, then we probably should do this change sooner than later.